### PR TITLE
Donation

### PR DESCRIPTION
**Pull request by babygoat**

### CODE REVIEW

**Comment by babygoat** [comment]
Re-submit to `feature-donations` branch

File: controllers/donation.go [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+			return 0, gin.H{}, models.NewAppError(errorWhere, "Fails to create a draft prime record", appErr.Error(), appErr.StatusCode)
+		default:
+			return http.StatusInternalServerError, gin.H{"status": "error", "message": fmt.Sprintf("Unknown Error Type. Fails to create a draft prime record. %v", err.Error())}, nil
+		}
+	}
+
+	tapPayReqJson, _ := json.Marshal(tapPayReq)
+
+	// Update the transaction status to 'paying'
+	mc.Storage.UpdateTransactionStatus(tapPayReq.OrderNumber, statusPaying, primeTableName)
```
Comment (nickhsine) [inline controllers/donation.go]: no error handling here?

File: controllers/donation.go [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+				Status:           statusError,
+			})
+		}
+		return 0, gin.H{}, models.NewAppError(errorWhere, err.Error(), "", http.StatusInternalServerError)
+	}
+
+	// Update the transaction status to 'paid' if transaction succeeds
+	updateRecord := buildPrimeSuccessRecord(tapPayResp)
+
+	mc.Storage.UpdateAPayByPrimeDonation(tapPayReq.OrderNumber, updateRecord)
```
Comment (nickhsine) [inline controllers/donation.go]: no error handling here?

File: controllers/donation.go [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+	tapPayReqJson, _ := json.Marshal(tapPayReq)
+
+	// Update the transaction status to 'paying'
+	mc.Storage.UpdateTransactionStatus(tapPayReq.OrderNumber, statusPaying, primeTableName)
+
+	tapPayResp, err := serveHttp(tapPayReq.PartnerKey, tapPayReqJson)
+
+	if nil != err {
+		if (tapPayTransactionResp{}) != tapPayResp {
```
Comment (nickhsine) [inline controllers/donation.go]: why do we need this condition?

File: storage/donation.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"bytes"
+
+	log "github.com/Sirupsen/logrus"
+
+	"twreporter.org/go-api/models"
+)
+
+// CreateAPayByPrimeDonation creates a draft order records in database
```
Comment (nickhsine) [inline storage/donation.go]: draft order ?

File: storage/donation.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+func (g *GormStorage) CreateAPayByPrimeDonation(m models.PayByPrimeDonation) error {
+	errWhere := "GormStorage.CreateAPayByPrimeDonation"
+	err := g.db.Create(&m).Error
+	if nil != err {
+		log.Error(err.Error())
+		return g.NewStorageError(err, errWhere, err.Error())
```
Comment (nickhsine) [inline storage/donation.go]: according to [this PR](https://github.com/twreporter/go-api/pull/138/files#diff-25bf9002ed579ca125ad5badacc2d822R45), the third argument is `message string`.
Hence, we may write code like
```
return g.NewStorageError(err, errWhere, fmt.Sprintf("can not create record(%#v)", m))
```

File: storage/donation.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+	var sqlStmt bytes.Buffer
+	sqlStmt.WriteString("UPDATE " + table + " ")
+	sqlStmt.WriteString("SET status='" + status + "' ")
+	sqlStmt.WriteString("WHERE order_number='" + order + "'")
+
+	err := g.db.Exec(sqlStmt.String()).Error
+	if nil != err {
+		log.Error(err.Error())
+		return g.NewStorageError(err, errWhere, err.Error())
```
Comment (nickhsine) [inline storage/donation.go]:
```
return g.NewStorageError(err, errWhere, fmt.Sprintf("can not update transaction status(order: %s, status: %s, table: %s)", order, status, table))
```

File: storage/donation.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	return nil
+}
+
+// UpdateAPayByPrimeDonation updates the draft record with the Tappay response data by order
+func (g *GormStorage) UpdateAPayByPrimeDonation(order string, m models.PayByPrimeDonation) error {
+	errWhere := "GormStorage.UpdateAPayByPrimeDonation"
+	err := g.db.Model(&m).Where("order_number = ?", order).Updates(m).Error
+	if nil != err {
+		log.Error(err.Error())
+		return g.NewStorageError(err, errWhere, err.Error())
```
Comment (nickhsine) [inline storage/donation.go]:
```
return g.NewStorageError(err, errWhere, fmt.Sprintf("can not update prime donation(order: %s, record: %#v)", order, m))
```

File: storage/donation.go (hunk 4 above)
Comment (babygoat) [reply storage/donation.go]: I will use singular word `record` in #139 .

File: controllers/donation.go (hunk 3 above)
Comment (babygoat) [reply controllers/donation.go]: Empty response implies that an error occurs during the tappay communication. This should be handled with the error handling cron job.

File: controllers/donation.go (hunk 1 above)
Comment (babygoat) [reply controllers/donation.go]: I will add the error logging here. However, I think the request should proceed even for the status update error.

File: controllers/donation.go (hunk 2 above)
Comment (babygoat) [reply controllers/donation.go]: I will add the error logging here. However, I think the request should proceed even for the status update error.

**Pull request closed without merging**